vault/ssorolecredentialsprovider.go

```diff
@@ -201,3 +211,191 @@ func (p *SSORoleCredentialsProvider) newOIDCToken(ctx context.Context) (*ssooidc
+	if err := open.Run(authorizeURL.String()); err != nil {
+		log.Printf("Failed to open browser: %s", err)
+	}
+
+	// await the authorization code from the callback server once the user has completed the flow.
+	var code string
+	timeout := time.After(1 * time.Minute)
+	select {
+	case <-timeout:
+		return nil, errors.New("timed out waiting for authorization code")
+	case code = <-cbServer.code:
+		log.Printf("Received authorization code: %s", code)
+		if err := cbServer.h.Close(); err != nil {
+			log.Printf("Failed to close oauthCallbackServer: %s", err)
+		}
+	}
+
+	// create the OIDC token using the authorization code received from the callback server
+	tok, err := p.OIDCClient.CreateToken(ctx, &ssooidc.CreateTokenInput{
+		ClientId:     clientCreds.ClientId,
+		ClientSecret: clientCreds.ClientSecret,
+		Code:         aws.String(code),
+		CodeVerifier: aws.String(codeVerifier),
+		GrantType:    aws.String("authorization_code"),
+		RedirectUri:  aws.String(redirectURI),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("Created new OIDC access token for %s (expires in: %ds)", p.StartURL, tok.ExpiresIn)
+	return tok, nil
+}
+
+// newOauthCallbackServer creates a HTTP server listening on a random localhost
+// port to serve the OAuth2 callback. It serves a single oauth callback endpoint
+// and sends the authorization code received via a channel.
+func newOauthCallbackServer() (*oauthCallbackServer, error) {
+	// select a random port for the callback server
+	ln, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create listener: %w", err)
+	}
+	log.Printf("oauthCallbackListener listening on %s", ln.Addr().String())
+
+	// create a 32 byte state for CSRF protection
+	state := make([]byte, 32)
+	n, err := crand.Read(state)
+	if err != nil || n != 32 {
+		return nil, fmt.Errorf("failed to generate state: %w", err)
+	}
+
+	oauth := &oauthCallbackServer{
+		state: base64.RawURLEncoding.EncodeToString(state),
+		code:  make(chan string),
+		ln:    ln,
+	}
+	oauth.h = &http.Server{
+		Handler: http.HandlerFunc(oauth.handleCallback),
+	}
+
+	return oauth, nil
+}
+
+// handleCallback handles the OAuth2 callback request and sends the authorization code to the server channel.
+func (s *oauthCallbackServer) handleCallback(w http.ResponseWriter, r *http.Request) {
+	// only respond to GET requests on the callback
+	if r.Method != http.MethodGet {
+		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
+		return
+	}
+	if r.URL.Path != "/oauth/callback" {
+		http.Error(w, "Not Found", http.StatusNotFound)
+		return
+	}
+
+	// constant time string comparison of want vs got state
+	state := r.URL.Query().Get("state")
+	if subtle.ConstantTimeCompare([]byte(state), []byte(s.state)) != 1 {
+		http.Error(w, "Invalid state", http.StatusBadRequest)
+		return
+	}
+
+	// send the authorization code to the channel
+	code := r.URL.Query().Get("code")
+	s.code <- code
```

Comment:
perhaps it's worth returning early if there's an error rather than letting the consumer timeout? In which case s.code could be a channel of a struct that can hold either an `error` or a `string`, and the consumer would do something like

```go
code, err := <- server.code
if err != nil {
   // error
}
```